### PR TITLE
CI: split benchmarks from other checks

### PR DESCRIPTION
**Pull request by afeld**
That job was the longest-running, so this should parallelize the various code checks with the benchmarks to get done in less wall clock time.

This pull request builds on https://github.com/pandas-dev/pandas/pull/39745.

cc @fangchenli

- [ ] ~~closes #xxxx~~
- [ ] tests added / passed
- [ ] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [ ] ~~whatsnew entry~~

### CODE REVIEW

File: .github/actions/setup/action.yml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+name: Set up pandas
```
Comment (jreback) [inline .github/actions/setup/action.yml:1]: can this be in more of the builds? (you change a few i c); normally do this as a separate PR (e.g. when you refactor want to do one thing), but ok to leave.

**Review by jreback — CHANGES_REQUESTED** [review]
not sure this is actually running we had 19 checks before, should show up in the list.

File: .github/actions/setup/action.yml (hunk 1 above)
Comment (afeld) [reply .github/actions/setup/action.yml:1]: > can this be in more of the builds?

Maybe! The environment gets set up various ways for various builds. I wasn't sure why, so just changed the ones that looked the same.

> normally do this as a separate PR

The refactoring was in service of splitting the two jobs, but noted, thanks.

**Comment by afeld** [comment]
> not sure this is actually running we had 19 checks before

Ah, whoops, I thought that was a GitHub Actions glitch, but as usual, Problem Exists Between Chair and Keyboard 😛  Hadn't realized that when using [an action in the same repository](https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#example-using-action-in-the-same-repository-as-the-workflow):

- The path needs to start with `./`
- The job needs to run `actions/checkout` before that action can be used (which makes sense, now that I think about it)
- Need to specify the `shell` for every step (`defaults` aren't inherited)

20 checks running now, as intended!

**Comment by jreback** [comment]
@afeld can you merge master

**Comment by jreback** [comment]
can you merge master

**Comment by jreback** [comment]
i merged this one: https://github.com/pandas-dev/pandas/pull/39745

**Comment by afeld** [comment]
Hmm, went from the longest run being `CI / Checks` at [48 minutes](https://github.com/pandas-dev/pandas/runs/1884091523?check_suite_focus=true) to `CI / Benchmarks` taking 46. Better, but not a _lot_ better 😕

Flaky tests failing, but `CI` ones are 🟢.

**Comment by jbrockmendel** [comment]
@afeld is this ready?

**Comment by datapythonista** [comment]
> Not sure if this is worth. We've got an overheat of 15 minutes, and we are saving 5 from 50 for an extra build. No strong opinion, but I think our CI should be faster without this with the amount of builds we usually have running (unless we've got many more than the 20 GH actions jobs, that I don't think we did).

@afeld thoughts on this?

**Comment by simonjayhawkins** [comment]
@afeld closing as stale. feel free to reopen if you want to continue with this. (@datapythonista has a question)
